### PR TITLE
Add analysis view table containing workflow timing statistics

### PR DESCRIPTION
**Pull request by JAllen42**
Partially addresses #1170
Linked with cylc/cylc-uiserver#434
Based on the proof of concept #1171

This adds a table view containing statistics of the timings of tasks in a workflow. Currently this includes mean, standard deviation, minimum, maximum, median and quartiles, which can be calculated using the queue times for tasks, run times or the total time. These can be filtered based on the name of the task and which platform the tasks was run on.

image

The proof of concept query which loads job information from the workflow database has been changed to load task information instead, and the subscription has been deleted and replaced with a button to rerun the query.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Review by oliver-sanders — COMMENTED** [review]
Looks good :+1:

**Comment by oliver-sanders** [comment]
This branch is now a little behind master and will require rebasing at some point:

```console
$ git fetch upstream
$ git checkout analysis_view
$ git branch analysis_view.safe  # make a safe copy before starting
$ git rebase upstream/master
$ # fix any conflicts flagged during rebase
$ git add -u && git rebase --continue  # if there were any conflicts
$ git push -f origin HEAD
```

**Comment by JAllen42** [comment]
This has been rebased and restructured slightly to make it easier to add the tests.

There might need to be a couple of minor changes if the structure of the data is tweaked in the sibling ui-server branch, but other than that I think this is ready for review

**Review by MetRonnie — COMMENTED** [review]
Looks good! I still need to test with a real workflow

@JAllen42 You need to add an entry to `.mailmap`, something like
```
Jamie Allen       <your.metoffice@email.address>      JAllen42     <your.git.commit@email.address>

File: src/views/Analysis.vue [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<template>
+  <div class="c-analysis" style="width: 100%; height: 100%">
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/views/Analysis.vue:19]: I think it's safer to specify than the rely on the contained elements pushing right to the edge. E.G. consider "background-color" or "@click" events.

File: src/components/cylc/analysis/AnalysisTable.vue [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          :items="tasks"
+          :sort-by.sync="sortBy"
+          dense
+          :footer-props="{
+            itemsPerPageOptions: [10, 20, 50, 100, 200, -1],
+            showFirstLastPage: true
+          }"
+          :options="{ itemsPerPage: 50 }"
+        >
+          <template
```
[the remark this answers is not in the document]
Comment (JAllen42) [reply src/components/cylc/analysis/AnalysisTable.vue:45]: Yes, good point

File: src/components/cylc/analysis/filter.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Function to find the unique platforms in an array of tasks
+ *
+ * @export
+ * @param {array} tasks - The tasks to search for unique platforms
+ * @return {array} - An array of unique platform objects
+ */
+export function platformOptions (tasks) {
+  const platformOptions = [{ text: 'All', value: null }]
+  const platforms = []
+  for (const task of Object.values(tasks)) {
```
[the remark this answers is not in the document]
Comment (JAllen42) [reply src/components/cylc/analysis/filter.js]: I thought I had tried that, but good to know that that does work

File: src/views/Analysis.vue [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+import {
+  matchTask,
+  platformOptions
+} from '@/components/cylc/analysis/filter'
+import {
+  mdiChartLine,
+  mdiRefresh
+} from '@mdi/js'
+
+// list of fields to request for tasks
```
[the remark this answers is not in the document]
Comment (JAllen42) [reply src/views/Analysis.vue]: Thanks, another thing that's good to know about

File: src/views/Analysis.vue [hunk 5]
```diff
@@ -245,9 +237,7 @@ export default {
   },
 
   computed: {
-    // a list of the workflow IDs this view is "viewing"
-    // NOTE: we plan multi-workflow functionality so we are writing views
-    // to be mult-workflow compatible in advance of this feature arriving
```
Comment (MetRonnie) [inline src/views/Analysis.vue]: I think this comment should be kept actually

File: src/views/Analysis.vue [hunk 6]
```diff
@@ -260,8 +250,6 @@ export default {
   },
 
   methods: {
-    // run the one-off query for historical job data and pass its results
-    // through the callback
```
Comment (MetRonnie) [inline src/views/Analysis.vue]: And this one probably

File: tests/unit/components/cylc/analysis/analysis.data.js
Comment (MetRonnie) [inline tests/unit/components/cylc/analysis/analysis.data.js:1]: Can delete this now

File: .mailmap [hunk 7]
```diff
@@ -20,3 +20,4 @@ Renovate Bot[bot]     <bot@renovateapp.com>
 github-actions[bot]   <github-actions@no-reply.github.com>                             <41898282+github-actions[bot]@users.noreply.github.com>
 Carol Barno           <carol.barno.dev@gmail.com>                                      <32536886+CarolBarno@users.noreply.github.com>
 Aaron Cole            <aaron@epiphron.co.nz>
+Jamie Allen           <jamie.allen@metoffice.gov.uk>           JAllen42                <69790602+JAllen42@users.noreply.github.com>
```
Comment (MetRonnie) [inline .mailmap:23]: Hmm, this didn't work. Maybe:
```suggestion
Jamie Allen           <jrallen42@gmail.com>           JAllen42
```

Comment (JAllen42) [reply .mailmap]: I've updated the commit email address, so hopefully that works now

**Review by MetRonnie — APPROVED** [review]
🚀

**Review by oliver-sanders — APPROVED** [review]
Looks good, works smoothly, nice!

Pagination, filtering, timing and platform selection all work well.

I tried being an impatient user hammering the refresh button and managed to find one small issue where repeat presses can cause duplicate tasks.

Press once:

Screenshot from 2023-04-14 14-42-57

Press three times:

Screenshot from 2023-04-14 14-43-20

There's a neat trick to handle users hammering buttons called `debounce` which you can find in the `lodash` library.

There's an example usage in [this PR](https://github.com/cylc/cylc-ui/pull/1268/files#diff-cba9714d6d5b082fd0cf11c0b5884b2a96d8dcc5bb2c112e911a9b3d2b85a1de) in `src/views/Log.vue`. You wrap your function with `debounce`, then duplicate calls within the specified timeframe will effectively be ignored:

```js
methods: {
  historicalQuery: debounce(
    async function () {
      this.tasks = []
      ...
    },
    500  // only call this once every 500ms
  }
}
```

This article provides an explanation: https://css-tricks.com/debouncing-throttling-explained-examples/#aa-debounce

This is relatively minor so not a blocker, but if possible it would be good to get this debounced befure it shows up in the wild.

**Review by hjoliver — APPROVED** [review]
Functional review: an excellent addition to Cylc - thanks @JAllen42 and @ChrisPaulBennett 

(I'll hold off merging till the debounce question is addressed).

**Comment by oliver-sanders** [comment]
I'm afraid you've picked up a couple more conflicts, [rebase instructions](https://github.com/cylc/cylc-ui/pull/1254#issuecomment-1453297280).

**Comment by oliver-sanders** [comment]
Cheers!

**Pull request merged**